src/codeu/chat/client/simplegui/ConversationPanel.java

```diff
@@ -121,19 +130,91 @@ public void actionPerformed(ActionEvent e) {
       @Override
       public void actionPerformed(ActionEvent e) {
         if (clientContext.user.hasCurrent()) {
-          final String s = (String) JOptionPane.showInputDialog(
-              ConversationPanel.this, "Enter title:", "Add Conversation", JOptionPane.PLAIN_MESSAGE,
-              null, null, "");
+          final String s = (String) JOptionPane.showInputDialog( ConversationPanel.this, 
+              "Enter title:", "Add Conversation", JOptionPane.PLAIN_MESSAGE, null, null, "");
+
           if (s != null && s.length() > 0) {
-            clientContext.conversation.startConversation(s, clientContext.user.getCurrent().id);
-            ConversationPanel.this.getAllConversations(listModel);
+            if (!clientContext.conversation.titleExists(s)) {
+
+              clientContext.conversation.startConversation(s, clientContext.user.getCurrent().id);
+
+              ConversationPanel.this.getAllConversations(listModel);
+            } else {
+
+              JOptionPane.showMessageDialog(ConversationPanel.this, "A conversation already exists with that name!",
+                "Error", JOptionPane.ERROR_MESSAGE);
+
+              actionPerformed(e);
+            }
           }
         } else {
           JOptionPane.showMessageDialog(ConversationPanel.this, "You are not signed in.");
         }
       }
     });
 
+    // User clicks Conversations Rename button.
+    renameButton.addActionListener(new ActionListener() {
+      @Override
+      public void actionPerformed(ActionEvent e) {
+        if (objectList.getSelectedIndex() != -1) {
+          final int index = objectList.getSelectedIndex();
+          final String data = objectList.getSelectedValue();
+          final ConversationSummary cs = ConversationPanel.this.lookupByTitle(data, index);
+
+          final String title = JOptionPane.showInputDialog(ConversationPanel.this, "Enter new title:",
+            "Rename Conversation", JOptionPane.PLAIN_MESSAGE);
+
+          final String currentTitle = cs.getTitle();
+
+          if (title != null) {
+            if (title.length() != 0 && !clientContext.conversation.titleExists(title)) {
```

Comment:
I feel like one of these checks is redundant. Either we can call something like Conversation.titleIsValid() which will return true/false, and you display messages based on that, or we assume that conversation.rename will not be called with invalid title